### ISSUE TITLE
NaN output from model with some weights

### ISSUE THREAD

**Issue by melihyilmaz** [Noble-Lab/casanovo#186]
We’ve identified some sets of weights for which the model outputs a NaN score tensor (this is the raw output of Casanovo decoder) at the very first step of decoding (output of [decoder](https://github.com/Noble-Lab/casanovo/blob/effc955ab488212f5f95ae02a896c66478535704/casanovo/denovo/model.py#L245)), hence leading to missing PTMs in the mztab output and possibly NaN loss during training. 

I can't add mgf and ckpt files here but a particular set of weights and spectrum that produce the NaN output are referenced in [this](https://noblelab.slack.com/archives/C01MXN4NWMP/p1682641415039889) Slack post. More details to come.

**Comment by wsnoble** [comment]
The files referenced above are too big to upload, but you can see them here:

[missing_spectra.mgf](https://noble.gs.washington.edu/~wnoble/temp/missing_spectra.mgf)

[2M_step=7000000.ckpt](https://noble.gs.washington.edu/~wnoble/temp/2M_step=700000.ckpt)

**Comment by melihyilmaz** [comment]
NaNs first appear in `memories`, i.e. spectrum embeddings, [here](https://github.com/Noble-Lab/casanovo/blob/6299bd26b44d0d4c9b3c0bd20d23bcf595a4b0c2/casanovo/denovo/model.py#L226), which is the output of [`SpectrumEncoder`](https://github.com/wfondrie/depthcharge/blob/ddbc8a6cdb6ce6c8e3478b6a4e9f29f124e8fced/depthcharge/components/transformers.py#L11) in DepthCharge. This means only spectrum peaks themselves, not the precursor info, leads to NaNs for this set of weights.

I also experimented with knocking out some peaks and discovered that some single peak knockouts change output of the spectrum encoder from all NaNs to a valid one and we get the correct peptide prediction in the mztab. I've found that knocking out peaks with <=340.14798 m/z doesn't lead to any NaNs while knockouts larger than this value still end up in NaNs. You can see a list of peaks I've individually knocked out and the outcome (+: correct prediction, -: NaN output) below:

101.07108 1439.6 +
197.12851 19921.1 +
312.15414 6194.9 +
326.16864 2459.0 +
330.16531 2423.6 +
340.14798 4524.4 +
344.18265 1238.4 - 
354.16476 15248.2 -
372.17520 8336.5 - 
451.18591 1865.8 - 
807.47083 32895.3 -
1343.70520 5006.4 - 

I don't have any insights about the significance of 340.14798 m/z and for reference this problematic spectrum satisfy all the constraints of our pre-processing filter (# peaks, max/min m/z etc.)

**Comment by melihyilmaz** [comment]
As I described above spectrum encoder produces NaN for certain set of weights and for spectra. Diving deeper, I’ve found that NaN peak embeddings first emerge as the output of the 3rd transformer encoder layer (20/126 spectrum peaks are NaNs) and everything becomes NaN the next layer.

Checking out the model weights, there are no NaNs and I couldn’t find a meaningful difference between 3rd layer and other layers (min,max,mean weight values etc.). I can look into the outputs of subcomponents the 3rd layer (self attention, linear etc) but let me know if you have ideas for any other way of probing this.

I’ve found some pytorch issues that seemed relevant at first but couldn’t glean much from them ([here](https://github.com/pytorch/pytorch/issues/41508) and [here](https://github.com/pytorch/pytorch/issues/24816)).

**Comment by melihyilmaz** [comment]
When I was trying to produce a minimal example on Google Colab for convenience, I've realized that the same set up described above that produces NaN (same model weights and spectrum), doesn't yield NaNs and returns the correct peptide identification. Editable Colab notebook is [here](https://colab.research.google.com/drive/1qKAs5KMYxxxb45CIrjY4nHdDx88oTWaI?usp=sharing), I'll investigate this further.

**Comment by melihyilmaz** [comment]
I resolved the issue* and the same Colab notebook linked above has a minimal example of a single spectrum resulting in NaN embeddings.

*: it took me a while to figure out but omission of `with torch.no_grad():` when encoding the spectrum peaks results in no NaNs. This must be implemented in the Pytorch Lightning under the hood as it's common to do so for all inference steps outside of training loops. Even after some googling, I don't understand why this changes the behavior.

**Comment by cfmelend** [comment]
I have been able to replicate @melihyilmaz Colab notebook without yielding nan's using the latest torch nightly (CUDA 12.1) build in combination with the environment (and install procedure) that @bittremieux suggested. I've tested this on: GTX 1080, RTX 2080, RTX 2080Ti, and A100 and successfully avoided the nan outputs in .eval() mode every time. I've attached my env to this post. The only idiosyncratic aspects to setting up the env were:

(1) using python 3.11.4 when creating the conda env

(2) installing jupyterlab and then using a jupyternotebook to run the install commands that are used in the Colab notebook (!pip install git+https://github.com/Noble-Lab/casanovo.git@6299bd2; !pip install 'depthcharge-ms==0.2.0' --force-reinstall)

(3) having to force-reinstall the newest version of typing-extensions + force-reinstalling the newest torch nightly versionOtherwise the procedure just involved copying the cells in the Colab notebook. The tensor outputs look very similar to those that were produced with a generic custom TransformerEncoder, with grad set to 'True', and with only cpu().
[nightly_env.txt](https://github.com/Noble-Lab/casanovo/files/12234564/nightly_env.txt)
